src/position-observer.js

```diff
@@ -0,0 +1,107 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {SubscriptionApi} from './iframe-helper';
+import {MessageType} from './3p-frame-messaging.js';
+import {rateLimit} from './utils/rate-limit';
+import {LayoutRectDef} from './layout-rect';
+
+/** @const */
+const MIN_EVENT_INTERVAL_IN_MS = 100;
+
+/**
+ * @typedef {{
+ *   viewport: !LayoutRectDef,
+ *   target: !LayoutRectDef
+ * }}
+ */
+let PositionEntryDef;
+
+export class PositionObserverApi {
```

Comment:
As you also pointed out `PositionObserver` is not a good name. I name it `PositionObserverApi` because we have a `IntersectionObserverApi` class, and I just name it blindly : )